File: frontend/src/api/scraping.ts
/**
 * API Service - Scraping & Sync
 * Web scraping control and Google Sheets synchronization
 */
import api from "../lib/axios";

/**
 * Status de uma tarefa de scraping
 */
export interface ScrapingStatus {
  task_id: string;
  status: "pending" | "running" | "completed" | "failed";
  progress: number; // 0-100
  current_step?: string;
  total_items?: number;
  processed_items?: number;
  errors?: string[];
  started_at?: string;
  completed_at?: string;
}

/**
 * Resultado da sincronização
 */
export interface SyncResult {
  success: boolean;
  message: string;
  records_created?: number;
  records_updated?: number;
  records_failed?: number;
  total_records?: number;
  errors?: string[];
}

/**
 * Inicia scraping de fotos do Transfermarkt
 */
export async function startPhotoScraping(): Promise<{ task_id: string }> {
  const { data } = await api.post<{ task_id: string }>("/scraping/photos/start");
  return data;
}

/**
 * Inicia scraping de dados gerais (nomes, clubes, etc.)
 */
export async function startDataScraping(): Promise<{ task_id: string }> {
  const { data } = await api.post<{ task_id: string }>("/scraping/data/start");
  return data;
}

/**
 * Busca status de uma tarefa de scraping
 */
export async function getScrapingStatus(taskId: string): Promise<ScrapingStatus> {
  const { data } = await api.get<ScrapingStatus>(`/scraping/status/${taskId}`);
  return data;
}

/**
 * Cancela tarefa de scraping em andamento
 */
export async function cancelScraping(taskId: string): Promise<void> {
  await api.post(`/scraping/cancel/${taskId}`);
}

/**
 * Sincroniza jogadores do Google Sheets
 */
export async function syncGoogleSheets(): Promise<SyncResult> {
  const { data } = await api.post<SyncResult>("/sync/google-sheets");
  return data;
}

/**
 * Exporta jogadores para Google Sheets
 */
export async function exportToGoogleSheets(): Promise<SyncResult> {
  const { data } = await api.post<SyncResult>("/sync/export-to-sheets");
  return data;
}

/**
 * Busca histórico de sincronizações
 */
export async function getSyncHistory(limit: number = 10): Promise<SyncResult[]> {
  const { data } = await api.get<SyncResult[]>("/sync/history", {
    params: { limit },
  });
  return data;
}

File: backend/app/api/v1/endpoints/scraping.py
"""
API Endpoints - Scraping
Web scraping control with BackgroundTasks for async processing
"""
from fastapi import APIRouter, BackgroundTasks, HTTPException, Depends
from sqlalchemy.orm import Session
from typing import Dict, Any
import uuid
from datetime import datetime

from app.api import deps
from app.core.database import get_database

router = APIRouter()

# In-memory storage for task status (in production, use Redis or database)
scraping_tasks: Dict[str, Dict[str, Any]] = {}


def update_task_status(
    task_id: str,
    status: str,
    progress: int = 0,
    current_step: str = None,
    processed_items: int = 0,
    total_items: int = 0,
    errors: list = None,
):
    """Update task status in memory"""
    if task_id not in scraping_tasks:
        scraping_tasks[task_id] = {
            "task_id": task_id,
            "started_at": datetime.utcnow().isoformat(),
        }

    scraping_tasks[task_id].update({
        "status": status,
        "progress": progress,
        "current_step": current_step,
        "processed_items": processed_items,
        "total_items": total_items,
        "errors": errors or [],
    })

    if status in ["completed", "failed"]:
        scraping_tasks[task_id]["completed_at"] = datetime.utcnow().isoformat()


def scrape_photos_task(task_id: str, db: Session):
    """
    Background task for photo scraping
    Simulates scraping process with progress updates
    """
    try:
        update_task_status(
            task_id,
            status="running",
            progress=0,
            current_step="Iniciando scraping de fotos...",
        )

        # TODO: Implement actual scraping logic
        # Example structure:
        # 1. Fetch all players without photos
        # 2. For each player, scrape Transfermarkt
        # 3. Download and save photo
        # 4. Update database

        # Simulated progress (replace with real logic)
        import time
        total_items = 100
        update_task_status(
            task_id,
            status="running",
            progress=10,
            current_step="Buscando jogadores sem fotos...",
            total_items=total_items,
        )
        time.sleep(2)

        for i in range(1, total_items + 1):
            # Simulate processing
            time.sleep(0.1)
            progress = int((i / total_items) * 100)
            update_task_status(
                task_id,
                status="running",
                progress=progress,
                current_step=f"Baixando foto {i}/{total_items}...",
                processed_items=i,
                total_items=total_items,
            )

        update_task_status(
            task_id,
            status="completed",
            progress=100,
            current_step="Scraping concluído!",
            processed_items=total_items,
            total_items=total_items,
        )

    except Exception as e:
        update_task_status(
            task_id,
            status="failed",
            progress=0,
            current_step="Erro ao executar scraping",
            errors=[str(e)],
        )


def scrape_data_task(task_id: str, db: Session):
    """
    Background task for data scraping
    Updates player information from Transfermarkt
    """
    try:
        update_task_status(
            task_id,
            status="running",
            progress=0,
            current_step="Iniciando scraping de dados...",
        )

        # TODO: Implement actual data scraping
        # Example structure:
        # 1. Fetch all players
        # 2. For each player, scrape updated data (name, club, age, etc.)
        # 3. Update database

        import time
        total_items = 50
        update_task_status(
            task_id,
            status="running",
            progress=10,
            current_step="Buscando jogadores...",
            total_items=total_items,
        )
        time.sleep(2)

        for i in range(1, total_items + 1):
            time.sleep(0.2)
            progress = int((i / total_items) * 100)
            update_task_status(
                task_id,
                status="running",
                progress=progress,
                current_step=f"Atualizando dados {i}/{total_items}...",
                processed_items=i,
                total_items=total_items,
            )

        update_task_status(
            task_id,
            status="completed",
            progress=100,
            current_step="Scraping de dados concluído!",
            processed_items=total_items,
            total_items=total_items,
        )

    except Exception as e:
        update_task_status(
            task_id,
            status="failed",
            progress=0,
            current_step="Erro ao executar scraping",
            errors=[str(e)],
        )


@router.post("/photos/start")
async def start_photo_scraping(
    background_tasks: BackgroundTasks,
    db: Session = Depends(get_database),
    current_user = Depends(deps.get_current_admin_user),
):
    """
    Start photo scraping task in background
    Requires admin privileges
    """
    task_id = str(uuid.uuid4())

    # Initialize task
    scraping_tasks[task_id] = {
        "task_id": task_id,
        "status": "pending",
        "progress": 0,
        "started_at": datetime.utcnow().isoformat(),
    }

    # Add background task
    background_tasks.add_task(scrape_photos_task, task_id, db)

    return {"task_id": task_id}


@router.post("/data/start")
async def start_data_scraping(
    background_tasks: BackgroundTasks,
    db: Session = Depends(get_database),
    current_user = Depends(deps.get_current_admin_user),
):
    """
    Start data scraping task in background
    Requires admin privileges
    """
    task_id = str(uuid.uuid4())

    scraping_tasks[task_id] = {
        "task_id": task_id,
        "status": "pending",
        "progress": 0,
        "started_at": datetime.utcnow().isoformat(),
    }

    background_tasks.add_task(scrape_data_task, task_id, db)

    return {"task_id": task_id}


@router.get("/status/{task_id}")
async def get_scraping_status(
    task_id: str,
    current_user = Depends(deps.get_current_active_user),
):
    """
    Get status of scraping task
    """
    if task_id not in scraping_tasks:
        raise HTTPException(status_code=404, detail="Task not found")

    return scraping_tasks[task_id]


@router.post("/cancel/{task_id}")
async def cancel_scraping(
    task_id: str,
    current_user = Depends(deps.get_current_admin_user),
):
    """
    Cancel scraping task
    Note: This is a soft cancel, task will stop at next checkpoint
    """
    if task_id not in scraping_tasks:
        raise HTTPException(status_code=404, detail="Task not found")

    update_task_status(
        task_id,
        status="cancelled",
        progress=scraping_tasks[task_id].get("progress", 0),
        current_step="Cancelado pelo usuário",
    )

    return {"message": "Task cancelled"}
